File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.java
import java.util.*;
class Solution {
    
    public String solution(String p) {
        String answer = "";
        if(isRight(p)) return p; // 올바른 괄호
        //올바른 괄호 아닐때 
        answer = transform(p);
        return answer;
    }
    boolean isRight(String s) {
        char[] c = s.toCharArray();
        Stack<Character> stack = new Stack<>();
        for(int i = 0; i < c.length; i++) {
            if(c[i] == '(') {
                stack.push(c[i]);
            }
            else if(c[i] == ')') {
                if(!stack.isEmpty() && stack.peek() == '(') {
                    stack.pop();
                }
            }
        }
        return stack.isEmpty();
    }
    String transform(String s) {
        if(s.length() == 0) return "";
        int check = 0;
        int index = 0;
        char[] c = s.toCharArray();
        for(int i = 0; i < c.length; i++) {
            if(c[i] == '(') check++;
            else if(c[i] ==')') check--;
            if(check == 0) {
                index = i+1;
                break;
            }
        }
        String u = s.substring(0,index);
        String v = s.substring(index);
        if(isRight(u)) {
            return u + transform(v);
        }
        else {
            String str = "(";
            str += transform(v);
            str += ")";
            u = u.substring(1,u.length()-1);
            for(char ch : u.toCharArray()) {
                if(ch == ')') str += '(';
                else str += ')';
            }
            return str;
        
        }
    }
}

File: 프로그래머스/2/60058. 괄호 변환/괄호 변환.js
function seperate(str) {
    let cnt1 = 0, cnt2 = 0;
   
    for (let i = 0; i < str.length; i++) {
        const c = str[i];
        if (c === '(') cnt1++;
        else cnt2++;
        if (cnt1 === cnt2) return i;
    }
    return str.length - 1;
}

function isCorrect(str) {
    const stack = [];
     for (let i = 0; i < str.length; i++) {
         if (str[i] === '(') stack.push(str[i]);
         else {
             if (stack.length === 0) return false;
             stack.pop();
         }
    }
    return stack.length === 0;
}

function reverse(str) {
    let s = '';
    for (let i = 0; i < str.length; i++) {
        if (str[i] === '(') s += ')'
        else s += '(';
    }
    return s;
}

function solution(p) {
    var answer = '';
    if (p === '') return '';
    
    const index = seperate(p);
    const u = p.slice(0,index+1);
    const v = p.slice(index+1);
    
    if (isCorrect(u)) {
        return u + solution(v);
    } else {
        return '(' + solution(v) + ')' + reverse(u.slice(1,-1));
    }
    return answer;
}